runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt

```diff
@@ -5,131 +5,155 @@
-
-        val headerName = "x-amz-checksum-$checksumAlgorithmName".lowercase()
-        logger.debug { "Resolved checksum header name: $headerName" }
-
-        // remove all checksum headers except for $headerName
-        // this handles the case where a user inputs a precalculated checksum, but it doesn't match the input checksum algorithm
-        req.headers.removeAllChecksumHeadersExcept(headerName)
-
-        val checksumAlgorithm = checksumAlgorithmName?.toHashFunction() ?: throw ClientException("Could not parse checksum algorithm $checksumAlgorithmName")
-
-        if (!checksumAlgorithm.isSupported) {
-            throw ClientException("Checksum algorithm $checksumAlgorithmName is not supported for flexible checksums")
-        }
-
-        if (req.body.isEligibleForAwsChunkedStreaming) {
-            req.header("x-amz-trailer", headerName)
-
-            val deferredChecksum = CompletableDeferred<String>(context.executionContext.coroutineContext.job)
-
-            if (req.headers[headerName] != null) {
-                logger.debug { "User supplied a checksum, skipping asynchronous calculation" }
-
-                val checksum = req.headers[headerName]!!
-                req.headers.remove(headerName) // remove the checksum header because it will be sent as a trailing header
-
-                deferredChecksum.complete(checksum)
-            } else {
-                logger.debug { "Calculating checksum asynchronously" }
-                req.body = req.body
-                    .toHashingBody(checksumAlgorithm, req.body.contentLength)
+        checksumAlgorithm(
+            requestChecksumRequired,
+            requestChecksumCalculation,
+            requestChecksumAlgorithm,
+            context,
+        )?.let { checksumAlgorithm ->
+            if (context.protocolRequest.body.isEligibleForAwsChunkedStreaming) { // Handle checksum calculation here
+                logger.debug { "Calculating checksum during transmission using: ${checksumAlgorithm::class.simpleName}" }
+
+                val request = context.protocolRequest.toBuilder()
+                val deferredChecksum = CompletableDeferred<String>(context.executionContext.coroutineContext.job)
+                val checksumHeader = checksumAlgorithmHeader(checksumAlgorithm)
+
+                request.body = request.body
+                    .toHashingBody(checksumAlgorithm, request.body.contentLength)
                     .toCompletingBody(deferredChecksum)
-            }
 
-            req.trailingHeaders.append(headerName, deferredChecksum)
-            return req.build()
-        } else {
-            return super.modifyBeforeSigning(context)
+                request.headers.append("x-amz-trailer", checksumHeader)
+                request.trailingHeaders.append(checksumHeader, deferredChecksum)
+                context.executionContext.emitBusinessMetric(checksumAlgorithm.toBusinessMetric())
+            } else { // Delegate checksum calculation to super class
```

Comment:
> // Delegate checksum calculation to super class

This comment is a little misleading, it's not actually going to delegate to the super class, it's going to use the `calculateChecksum` and `applyChecksum` functions you defined below.

Reply:
What do you think would be a better comment?